File: CSharp/CodeCarvings.Piczard.Examples.ExampleSet_A/App_Code/MyRotationFilter.cs
/*
 * Piczard Examples | ExampleSet -A- C#
 * Copyright 2011-2012 Sergio Turolla - All Rights Reserved.
 * Author: Sergio Turolla
 * <codecarvings.com>
 *  
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF 
 * ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT 
 * SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR 
 * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN 
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
 * OR OTHER DEALINGS IN THE SOFTWARE.
 */

using System;
using System.Collections.Generic;
using System.Web;
using System.Drawing;

using CodeCarvings.Piczard;
using CodeCarvings.Piczard.Processing;
using CodeCarvings.Piczard.Helpers;
using CodeCarvings.Piczard.Serialization;

// Custom Rotation ImageProcessingFilter (Used by Example A.404)
// NOTE: This filter implements JSON serialziation
[Serializable]
public class MyRotationFilter
    : ImageProcessingFilter, IJSONSerializable
{

    #region Constructors

    public MyRotationFilter(float rotationAngle)
        : base()
    {
        this.RotationAngle = rotationAngle;
    }

    public MyRotationFilter()
        : this(0F)
    {
    }

    #endregion

    #region Overriedes

    protected override void Apply(ImageProcessingActionExecuteArgs args)
    {
        float normalizedRotationAngle = this._RotationAngle % 360F;
        if (normalizedRotationAngle == 0F)
        {
            // No need to rotate the image
            return;
        }

        Bitmap result = null;
        try
        {
            // Intial calculations
            double t;
            double t1;
            double b1;
            double t2;
            double b2;
            this.CalculateValues(args.Image.Size, out t, out t1, out b1, out t2, out b2);

            // Calculate the new image size
            Size outputImageSize = this.GetOutputImageSize(t1, b1, t2, b2);
           
            // Create the result image
            result = new Bitmap(outputImageSize.Width, outputImageSize.Height, CodeCarvings.Piczard.CommonData.DefaultPixelFormat);

            // Set the right image resolution (DPI)
            ImageHelper.SetImageResolution(result, args.ImageProcessingJob.OutputResolution);

            using (Graphics g = Graphics.FromImage(result))
            {
                // Use the max quality
                ImageHelper.SetGraphicsMaxQuality(g);

                if ((args.IsLastAction) && (!args.AppliedImageBackColorValue.HasValue))
                {
                    // Optimization (not mandatory)
                    // This is the last filter action and the ImageBackColor has not been yet applied...
                    // Apply the ImageBackColor now to save RAM & CPU
                    args.ApplyImageBackColor(g);
                }

                // Calculate the points for the DrawImage method
                Point[] shapePoints = new Point[3];

                if ((t >= 0D) && (t < (Math.PI / 2D)))
                {
                    shapePoints[0] = new Point(Convert.ToInt32(Math.Round(b2)), 0);
                    shapePoints[1] = new Point(outputImageSize.Width, Convert.ToInt32(Math.Round(t2)));
                    shapePoints[2] = new Point(0, Convert.ToInt32(Math.Round(b1)));
                }
                else if ((t >= (Math.PI / 2D)) && (t < Math.PI))
                {
                    shapePoints[0] = new Point(outputImageSize.Width, Convert.ToInt32(Math.Round(t2)));
                    shapePoints[1] = new Point(Convert.ToInt32(Math.Round(t1)), outputImageSize.Height);
                    shapePoints[2] = new Point(Convert.ToInt32(Math.Round(b2)), 0);
                }
                else if ((t >= Math.PI) && (t < (Math.PI + (Math.PI / 2D))))
                {
                    shapePoints[0] = new Point(Convert.ToInt32(Math.Round(t1)), outputImageSize.Height);
                    shapePoints[1] = new Point(0, Convert.ToInt32(Math.Round(b1)));
                    shapePoints[2] = new Point(outputImageSize.Width, Convert.ToInt32(Math.Round(t2)));
                }
                else
                {
                    shapePoints[0] = new Point(0, Convert.ToInt32(Math.Round(b1)));
                    shapePoints[1] = new Point(Convert.ToInt32(Math.Round(b2)), 0);
                    shapePoints[2] = new Point(Convert.ToInt32(Math.Round(t1)), outputImageSize.Height);
                }

                // Draw the image
                g.DrawImage(args.Image, shapePoints);
            }

            // Return the image
            args.Image = result;
        }
        catch
        {
            // An error has occurred...

            // Release the resources
            if (result != null)
            {
                result.Dispose();
                result = null;
            }

            // Re-throw the exception
            throw;
        }
    }

    #endregion

    #region Methods

    private void CalculateValues(Size imageSize, out double t, out double t1, out double b1, out double t2, out double b2)
    {
        // Convert image size to double
        double imageWidth = Convert.ToDouble(imageSize.Width);
        double imageHeight = Convert.ToDouble(imageSize.Height);

        // Convert the rotation angle
        t = Convert.ToDouble(this._RotationAngle) * Math.PI / 180D;
        while (t < 0D)
        {
            t += (Math.PI * 2D);
        }

        if (
            ((t >= 0D) && (t < (Math.PI / 2D)))
            ||
            ((t >= Math.PI) && (t < (Math.PI + (Math.PI / 2D)))))
        {
            t1 = Math.Abs(Math.Cos(t)) * imageWidth;
            b1 = Math.Abs(Math.Cos(t)) * imageHeight;

            t2 = Math.Abs(Math.Sin(t)) * imageWidth;
            b2 = Math.Abs(Math.Sin(t)) * imageHeight;
        }
        else
        {
            t1 = Math.Abs(Math.Sin(t)) * imageHeight;
            b1 = Math.Abs(Math.Sin(t)) * imageWidth;

            t2 = Math.Abs(Math.Cos(t)) * imageHeight;
            b2 = Math.Abs(Math.Cos(t)) * imageWidth;
        }
    }

    private Size GetOutputImageSize(double t1, double b1, double t2, double b2)
    {
        int width = Convert.ToInt32(Math.Ceiling(t1 + b2));
        int height = Convert.ToInt32(Math.Ceiling(b1 + t2));
        return new Size(width, height);
    }

    public Size GetOutputImageSize(Size imageSize)
    {
        double t;
        double t1;
        double b1;
        double t2;
        double b2;
        this.CalculateValues(imageSize, out t, out t1, out b1, out t2, out b2);

        return this.GetOutputImageSize(t1, b1, t2, b2);
    }

    #endregion

    #region Properties

    private float _RotationAngle;
    public float RotationAngle
    {
        get
        {
            return this._RotationAngle;
        }
        set
        {
            this._RotationAngle = value;
        }
    }

    #endregion

    #region Serialization

    int IJSONSerializable.SerializationVersion
    {
        get
        {
            // Default version = 1
            return 1;
        }
    }

    JSONSerializationException IJSONSerializable.GetSerializationException()
    {
        // No error
        return null;
    }

    JSONObject IJSONSerializable.ToJSONObject(JSONSerializationOptions options)
    {
        // Options cannot be null
        if (options == null)
        {
            options = JSONSerializationOptions.Default;
        }

        JSONObject result = options.GetNewJSONObject(this);
        result.SetNumberValue("rotationAngle", this._RotationAngle);

        return result;
    }

    private static MyRotationFilter FromJSON_1(JSONObject jsonObject)
    {
        if (jsonObject == null)
        {
            return null;
        }

        MyRotationFilter result = new MyRotationFilter();
        result.RotationAngle = jsonObject.GetNumberSingleValue("rotationAngle").Value;

        return result;
    }

    public static MyRotationFilter FromJSON(JSONObject jsonObject)
    {
        if (jsonObject == null)
        {
            return null;
        }

        int version = jsonObject.GetSerializationVersion();
        switch (version)
        {
            case 1:
                return FromJSON_1(jsonObject);
        }

        throw new JSONInvalidObjectException(typeof(MyRotationFilter));
    }

    public static MyRotationFilter FromJSON(string jsonString)
    {
        return FromJSON(JSONObject.Decode(jsonString));
    }

    #endregion

}

File: VB.NET/CodeCarvings.Piczard.Examples.ExampleSet_A/App_Code/MyRotationFilter.vb
' -------------------------------------------------------
' Piczard Examples | ExampleSet -A- VB.NET
' Copyright 2011-2013 Sergio Turolla - All Rights Reserved.
' Author: Sergio Turolla
' <codecarvings.com>
'  
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF 
' ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED 
' TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
' PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT 
' SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR 
' ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN 
' ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, 
' OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE 
' OR OTHER DEALINGS IN THE SOFTWARE.
' -------------------------------------------------------

Imports Microsoft.VisualBasic
Imports System.Drawing

Imports CodeCarvings.Piczard
Imports CodeCarvings.Piczard.Processing
Imports CodeCarvings.Piczard.Helpers
Imports CodeCarvings.Piczard.Serialization

' Custom Rotation ImageProcessingFilter (Used by Example A.404)
' NOTE: This filter implements JSON serialziation
<Serializable()> Public Class MyRotationFilter
    Inherits ImageProcessingFilter
    Implements IJSONSerializable

#Region "Constructors"

    Public Sub New(ByVal rotationAngle As Single)
        Me.RotationAngle = rotationAngle
    End Sub

    Public Sub New()
        Me.New(0.0F)

    End Sub

#End Region

#Region "Overriedes"

    Protected Overrides Sub Apply(ByVal args As CodeCarvings.Piczard.Processing.ImageProcessingActionExecuteArgs)
        Dim normalizedRotationAngle As Single = Me._RotationAngle Mod 360.0F
        If (normalizedRotationAngle = 0.0F) Then
            ' No need to rotate the image
            Return
        End If

        Dim result As Bitmap = Nothing
        Try
            ' Intial calculations
            Dim t As Double
            Dim t1 As Double
            Dim b1 As Double
            Dim t2 As Double
            Dim b2 As Double

            Me.CalculateValues(args.Image.Size, t, t1, b1, t2, b2)

            ' Calculate the new image size
            Dim outputImageSize As Size = Me.GetOutputImageSize(t1, b1, t2, b2)

            ' Create the result image
            result = New Bitmap(outputImageSize.Width, outputImageSize.Height, CodeCarvings.Piczard.CommonData.DefaultPixelFormat)

            ' Set the right image resolution (DPI)
            ImageHelper.SetImageResolution(result, args.ImageProcessingJob.OutputResolution)

            Using g As Graphics = Graphics.FromImage(result)
                ' Use the max quality
                ImageHelper.SetGraphicsMaxQuality(g)

                If ((args.IsLastAction) And (Not args.AppliedImageBackColorValue.HasValue)) Then
                    ' Optimization (not mandatory)
                    ' This is the last filter action and the ImageBackColor has not been yet applied...
                    ' Apply the ImageBackColor now to save RAM & CPU
                    args.ApplyImageBackColor(g)
                End If

                ' Calculate the points for the DrawImage method
                Dim shapePoints(2) As Point

                If ((t >= 0.0R) And (t < (Math.PI / 2.0R))) Then
                    shapePoints(0) = New Point(Convert.ToInt32(Math.Round(b2)), 0)
                    shapePoints(1) = New Point(outputImageSize.Width, Convert.ToInt32(Math.Round(t2)))
                    shapePoints(2) = New Point(0, Convert.ToInt32(Math.Round(b1)))
                ElseIf ((t >= (Math.PI / 2.0R)) And (t < Math.PI)) Then
                    shapePoints(0) = New Point(outputImageSize.Width, Convert.ToInt32(Math.Round(t2)))
                    shapePoints(1) = New Point(Convert.ToInt32(Math.Round(t1)), outputImageSize.Height)
                    shapePoints(2) = New Point(Convert.ToInt32(Math.Round(b2)), 0)
                ElseIf ((t >= Math.PI) And (t < (Math.PI + (Math.PI / 2.0R)))) Then
                    shapePoints(0) = New Point(Convert.ToInt32(Math.Round(t1)), outputImageSize.Height)
                    shapePoints(1) = New Point(0, Convert.ToInt32(Math.Round(b1)))
                    shapePoints(2) = New Point(outputImageSize.Width, Convert.ToInt32(Math.Round(t2)))
                Else
                    shapePoints(0) = New Point(0, Convert.ToInt32(Math.Round(b1)))
                    shapePoints(1) = New Point(Convert.ToInt32(Math.Round(b2)), 0)
                    shapePoints(2) = New Point(Convert.ToInt32(Math.Round(t1)), outputImageSize.Height)
                End If

                ' Draw the image
                g.DrawImage(args.Image, shapePoints)
            End Using

            ' Return the image
            args.Image = result
        Catch
            ' An error has occurred...

            ' Release the resources
            If (result IsNot Nothing) Then
                result.Dispose()
                result = Nothing
            End If

            ' Re-throw the exception
            Throw
        End Try
    End Sub

#End Region

#Region "Methods"

    Private Sub CalculateValues(ByVal imageSize As Size, ByRef t As Double, ByRef t1 As Double, ByRef b1 As Double, ByRef t2 As Double, ByRef b2 As Double)
        ' Convert image size to double
        Dim imageWidth As Double = Convert.ToDouble(imageSize.Width)
        Dim imageHeight As Double = Convert.ToDouble(imageSize.Height)

        ' Convert the rotation angle
        t = Convert.ToDouble(Me._RotationAngle) * Math.PI / 180.0R
        While (t < 0.0R)
            t += (Math.PI * 2.0R)
        End While

        If ( _
            ((t >= 0.0R) And (t < (Math.PI / 2.0R))) _
            Or _
            ((t >= Math.PI) And (t < (Math.PI + (Math.PI / 2.0R))))) Then
            t1 = Math.Abs(Math.Cos(t)) * imageWidth
            b1 = Math.Abs(Math.Cos(t)) * imageHeight

            t2 = Math.Abs(Math.Sin(t)) * imageWidth
            b2 = Math.Abs(Math.Sin(t)) * imageHeight
        Else
            t1 = Math.Abs(Math.Sin(t)) * imageHeight
            b1 = Math.Abs(Math.Sin(t)) * imageWidth

            t2 = Math.Abs(Math.Cos(t)) * imageHeight
            b2 = Math.Abs(Math.Cos(t)) * imageWidth
        End If
    End Sub

    Private Function GetOutputImageSize(ByVal t1 As Double, ByVal b1 As Double, ByVal t2 As Double, ByVal b2 As Double) As Size
        Dim width As Integer = Convert.ToInt32(Math.Ceiling(t1 + b2))
        Dim height As Integer = Convert.ToInt32(Math.Ceiling(b1 + t2))
        Return New Size(width, height)
    End Function

    Public Function GetOutputImageSize(ByVal imageSize As Size) As Size
        Dim t As Double
        Dim t1 As Double
        Dim b1 As Double
        Dim t2 As Double
        Dim b2 As Double
        Me.CalculateValues(imageSize, t, t1, b1, t2, b2)

        Return Me.GetOutputImageSize(t1, b1, t2, b2)
    End Function

#End Region

#Region "Properties"
    Private _RotationAngle As Single
    Public Property RotationAngle() As Single
        Get
            Return Me._RotationAngle
        End Get
        Set(ByVal value As Single)
            Me._RotationAngle = value
        End Set
    End Property
#End Region

#Region "Serialization"

    Protected ReadOnly Property SerializationVersion() As Integer Implements IJSONSerializable.SerializationVersion
        Get
            ' Default version = 1
            Return 1
        End Get
    End Property

    Protected Function GetSerializationException() As JSONSerializationException Implements IJSONSerializable.GetSerializationException
        ' No error
        Return Nothing
    End Function

    Protected Function ToJSONObject(ByVal options As JSONSerializationOptions) As JSONObject Implements IJSONSerializable.ToJSONObject
        ' Options cannot be Nothing
        If (options Is Nothing) Then
            options = JSONSerializationOptions.Default
        End If

        Dim result As JSONObject = options.GetNewJSONObject(Me)
        result.SetNumberValue("rotationAngle", Me._RotationAngle)

        Return result
    End Function

    Private Shared Function FromJSON_1(ByVal jsonObject As JSONObject) As MyRotationFilter
        If (jsonObject Is Nothing) Then
            Return Nothing
        End If

        Dim result As MyRotationFilter = New MyRotationFilter()
        result.RotationAngle = jsonObject.GetNumberSingleValue("rotationAngle").Value

        Return result
    End Function

    Public Shared Function FromJSON(ByVal jsonObject As JSONObject) As MyRotationFilter
        If (jsonObject Is Nothing) Then
            Return Nothing
        End If

        Dim version As Integer = jsonObject.GetSerializationVersion()
        Select Case version
            Case 1
                Return FromJSON_1(jsonObject)
        End Select

        Throw New JSONInvalidObjectException(GetType(MyRotationFilter))
    End Function

    Public Shared Function FromJSON(ByVal jsonString As String) As MyRotationFilter
        Return FromJSON(JSONObject.Decode(jsonString))
    End Function

#End Region

End Class
